File: movieland-backend/src/main/java/com/phil/movieland/rest/service/ReservationService.java
package com.phil.movieland.rest.service;

import com.phil.movieland.data.entity.Movie;
import com.phil.movieland.data.entity.MovieShow;
import com.phil.movieland.data.entity.Reservation;
import com.phil.movieland.data.entity.Seat;
import com.phil.movieland.data.repository.ReservationRepository;
import com.phil.movieland.data.repository.SeatRepository;
import com.phil.movieland.utils.Utils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Date;
import java.util.List;
import java.util.Optional;

@Service
public class ReservationService {

    private final SeatRepository seatRepository;
    private final ReservationRepository reservationRepository;

    private Logger log = LoggerFactory.getLogger(ReservationService.class);

    @Autowired
    public ReservationService(SeatRepository seatRepository, ReservationRepository reservationRepository) {
        this.seatRepository = seatRepository;
        this.reservationRepository = reservationRepository;
    }


    public Reservation updateReservation(Reservation reservation) {
        return reservationRepository.save(reservation);
    }

    public Reservation saveReservation(Reservation reservation, List<Seat> seats) {
        /** Check if seats are already taken*/
        //TOD Use seatRepository.findAllOfShow
        List<Reservation> reservations = reservationRepository.findAllByShowId(reservation.getShowId());
        for (Reservation res : reservations) {
            List<Seat> seatList = seatRepository.findAllByResId(res.getResId());
            for (Seat reserved : seatList) {
                if (seats.stream().anyMatch(selected -> reserved.getNumber() == selected.getNumber())) {
                    return null;
                }
            }
        }
        /**Calculate total Sum */
        double totalSum = seats.stream().mapToDouble(seat -> Seat.getPrice(seat.getType())).sum();
        reservation.setTotalSum(totalSum);

        Reservation result = reservationRepository.save(reservation);
        /** Store Seats */
        seats.stream().forEach(seat -> {
            seat.setResId(result.getResId());
            seatRepository.save(seat);
        });
        return result;
    }

    public List<Reservation> getAllReservations() {
        log.info("Querying all Reservations");
        return reservationRepository.findAll();

    }

    public List<Reservation> getAllReservationsOfShow(long showId) {
        log.info("Querying all Reservations of MovieShow '{}'", showId);
        return reservationRepository.findAllByShowId(showId);
    }

    public List<Seat> getAllSeatsOfReservation(Long resId) {
        log.info("Querying all Seats of Reservation '{}'", resId);
        return seatRepository.findAllByResId(resId);
    }

    public List<Seat> getAllSeatsOfShow(Long showId) {
        log.info("Querying all Seats of Show '{}'", showId);
        return seatRepository.findSeatsOfShow(showId);
    }

    public Optional<Reservation> getReservationById(Long resId) {
        log.info("Querying Reservation by id='{}'", resId);
        return reservationRepository.findById(resId);
    }


    public void deleteAll() {
        log.info("Deleting all Reservations");
        List<Reservation> reservations = reservationRepository.findAll();
        reservations.forEach(reservation -> seatRepository.deleteAllByResId(reservation.getResId()));
        reservationRepository.deleteAll();
    }

    public List<Reservation> getAllReservationsOfUser(Long userId, boolean futureReservations) {
        log.info("Querying all Reservations of User userId='{}'", userId);
        if (futureReservations) {
            return reservationRepository.findFutureAllByUserId(userId, new Date());
        }
        return reservationRepository.findAllByUserId(userId);
    }

    public void saveReservationsWithSeats(List<StatisticsService.ReservationWithSeats> reservations) {
        log.info("Saving Reservations List");
        reservations.forEach(res -> saveReservation(res.getReservation(), res.getSeats()));
    }

    /**
     * Return if seatList is available for show
     */
    //TODO REPLACE WITH SQL STATEMENT
    public boolean areSeatsAvailable(long showId, List<Seat> seatList) {
       /* List<Reservation> reservations= reservationRepository.findAllByShowId(showId);
        for(Reservation res: reservations){
            List<Seat> takenSeats=seatRepository.findAllByResId(res.getResId());
            for(Seat takenSeat: takenSeats){
                *//** Return false if takenSeat matches any in the list*//*
                if(seatList.stream().anyMatch(s-> s.getNumber()==takenSeat.getNumber()))
                    return false;
            }
        }
        return true;*/
        log.info("Check if Seats '{}' are available for MovieShow '{}'", Utils.joinToStringList(seatList), showId);
        for (Seat seat : seatList) {
            if (!seatRepository.findSeatDuplicates(seat.getNumber(), showId).isEmpty()) {
                log.info("Seat ({} for MovieShow: '{}') is already taken", seat.getNumber(), showId);
                return false;
            }
        }
        return true;
    }

    /**
     * Returns amount of deleted reservations
     */
    public long deleteReservationsOfShows(List<Long> showIds) {
        log.info("Deleting all Reservations of MovieShows showIds='{}'", Utils.joinToStringList(showIds));
        return reservationRepository.deleteAllByShowIdIn(showIds);
    }

    public Optional<Reservation> getReservationInfoOfUser(Long userId, Long resId) {
        log.info("Querying Reservation Info of User userId='{}', resId='{}'", userId, resId);
        return reservationRepository.findByResIdAndUserId(resId, userId);
    }

    public static class ReservationInfo {
        private Reservation reservation;
        private MovieShow movieShow;
        private Movie movie;
        private String QRCodeURL;

        public Reservation getReservation() {
            return reservation;
        }

        public void setReservation(Reservation reservation) {
            this.reservation = reservation;
        }

        public MovieShow getMovieShow() {
            return movieShow;
        }

        public void setMovieShow(MovieShow movieShow) {
            this.movieShow = movieShow;
        }

        public Movie getMovie() {
            return movie;
        }

        public void setMovie(Movie movie) {
            this.movie = movie;
        }

        public String getQRCodeURL() {
            return QRCodeURL;
        }

        public void setQRCodeURL(String QRCodeURL) {
            this.QRCodeURL = QRCodeURL;
        }
    }
}

File: movieland-backend/rest/service/reservation_service.py
from typing import List

from db.database import db
from db.model import MovieShow, Reservation, Seat
from logger import get_logger
from rest.service.dto.reservation_with_seats import ReservationWithSeats
from rest.service.seats_service import SeatsService

log = get_logger()
seats_service = SeatsService()


class ReservationService:

    def get_reservation_by_id(self, id: int) -> Reservation:
        return Reservation.query.filter_by(resId=id).first()

    def get_reservation_by_id_and_user(self, id: int, userId: int) -> Reservation:
        return Reservation.query.filter_by(resId=id, userId=userId).first()

    def get_reservations_by_show_id(self, showId: int) -> List[Reservation]:
        return Reservation.query.filter_by(showId=showId).all()

    def get_reservations_by_user(self, user_id: int) -> List[Reservation]:
        return Reservation.query.filter_by(userId=user_id).all()

    def get_all(self) -> List[Reservation]:
        return Reservation.query.all()

    def save_reservation(self, reservation: Reservation, seats: List[Seat] = None, update=False) -> Reservation:
        if seats is not None:
            # Check if seats are already taken
            reservations = self.get_reservations_by_show(reservation.showId)
            for res in reservations:
                res_seats = seats_service.get_seats_of_reservation(res.resId)
                already_taken_seats = set(res_seats).intersection(seats)
                if len(already_taken_seats) > 0:
                    raise FileExistsError(f"Seats {already_taken_seats} are already taken")

        # Calculate total sum
        if seats is not None:
            total_sum = sum([Seat.prices[seat.type.name] for seat in seats])
            reservation.totalSum = total_sum
        if update:
            db.session.merge(reservation)
        else:
            db.session.add(reservation)
        db.session.commit()
        if seats is not None:
            for seat in seats:
                seat.resId = reservation.resId
            seats_service.save_seats(seats)
        return reservation

    def delete_all(self):
        Reservation.query.delete()

    def delete_by_ids(self, res_ids: List[int]):
        Reservation.query.filter(Reservation.resId.in_(res_ids)).delete()

    def get_reservations_by_show(self, showId: int) -> List[Reservation]:
        return Reservation.query.filter_by(showId=showId).all()

    def get_seats_of_reservation(self, resId: int) -> List[Seat]:
        return seats_service.get_seats_of_reservation(resId)

    def get_seats_of_show(self, showId: int) -> List[Seat]:
        return Seat.query.join(Reservation).join(MovieShow).filter_by(showId=showId).all()

    def delete_reservations_of_shows(self, show_ids: List[int]):
        seats_service.delete_by_show_ids(show_ids)
        return Reservation.query.filter(Reservation.showId.in_(show_ids)).delete()

    def are_seats_available(self, showId: int, seat_list: List[Seat]) -> bool:
        for seat in seat_list:
            if not seats_service.is_seat_available(seat.number, showId):
                return False
        return True

    def save_reservations_with_seats(self, reservations: List[ReservationWithSeats]):
        log.info("Saving Reservations List")
        for reservation in reservations:
            self.save_reservation(reservation.reservation, reservation.seats)

    def get_reservations_by_shows(self, show_ids: List[int]) -> List[Reservation]:
        return Reservation.query.filter(Reservation.showId.in_(show_ids)).all()

    def get_seats_of_reservations(self, res_ids: List[int]):
        return Seat.query.filter(Seat.resId.in_(res_ids)).all()
